flexmock reports MethodSignatureError on mocked sys.stdout.write

Hi,

I'm the maintainer of flexmock for gentoo and received the following bug report:
https://bugs.gentoo.org/740128

The software [borgmatic](https://torsion.org/borgmatic/) uses flexmock for testing.

One of the tests mocks `sys.stdout` and throws the following error:
```
___________________________________________ test_collect_configuration_run_summary_logs_outputs_merged_json_results ___________________________________________

    def test_collect_configuration_run_summary_logs_outputs_merged_json_results():
        flexmock(module).should_receive('run_configuration').and_return(['foo', 'bar']).and_return(
            ['baz']
        )
>       flexmock(module.sys.stdout).should_receive('write').with_args('["foo", "bar", "baz"]').once()

tests/unit/commands/test_borgmatic.py:450:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/lib/python3.8/site-packages/flexmock.py:391: in with_args
    self._verify_signature_match(*kargs, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <flexmock.Expectation object at 0x7f0724955bb0>, kargs = ('["foo", "bar", "baz"]',), kwargs = {}, is_method = False, args_len = 2, minimum = 2
maximum = 2, total_positional = 1, named_optionals = []

    def _verify_signature_match(self, *kargs, **kwargs):
        if isinstance(self._mock, Mock):
            return  # no sense in enforcing this for fake objects
        allowed = self.argspec
        # we consider object a method for purposes or not counting "self"/"cls" as argument if:
        #  - one of inspect.ismethod, inspect.isfunction, _isclass return True
        #    (in Python 3 it's sometimes impossible to tell whether callable is method or not,
        #     so we try both inspect.ismethod and inspect.isfunction; classes are callable too -
        #     they have __init__)
        #  - it's not a static method
        #  - the mocked object is a module - module "methods" are in fact plain functions;
        #    unless they're classes, which means they still have __init__
        is_method = ((inspect.ismethod(self.original) or inspect.isfunction(self.original)
                      or _isclass(self.original)) and
                     self.method_type is not staticmethod and
                     (not isinstance(self._mock, types.ModuleType) or
                      _isclass(self.original)))
        args_len = len(allowed.args)
        if is_method:
            args_len -= 1
        minimum = args_len - (allowed.defaults and len(allowed.defaults) or 0)
        maximum = None
        if allowed.varargs is None and allowed.keywords is None:
            maximum = args_len
        total_positional = len(
            kargs + tuple(a for a in kwargs if a in allowed.args))
        named_optionals = [a for a in kwargs
                           if allowed.defaults
                           if a in allowed.args[len(allowed.args) - len(allowed.defaults):]]
        if allowed.defaults and total_positional == minimum and named_optionals:
            minimum += len(named_optionals)
        if total_positional < minimum:
>           raise MethodSignatureError(
                '%s requires at least %s arguments, expectation provided %s' %
                (self.name, minimum, total_positional))
E           flexmock.MethodSignatureError: write requires at least 2 arguments, expectation provided 1

/usr/lib/python3.8/site-packages/flexmock.py:278: MethodSignatureError

```

The test failure has been reproduced by me for python 3.6, 3.7 and 3.9.
The following is a minimal example:

```python
import sys
import flexmock

flexmock(sys.stdout).should_receive('write').with_args('test').once()
```

Interesting is that the maintainer of borgmatic is unable to reproduce the issue (see his comment in the bug https://bugs.gentoo.org/740128#c2).

--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -32,6 +32,10 @@ Removed
 - Drop Pytest 4.x support.
 - Remove unittest2 and nose integrations. unittest2 and nose are not maintained anymore.
 
+Fixed
+#####
+
+- Fix `with_args` not working built-in functions and methods.
 
 Infrastructure
 ##############

--- a/src/flexmock/api.py
+++ b/src/flexmock/api.py
@@ -4,7 +4,7 @@ import inspect
 import re
 import sys
 import types
-from types import TracebackType
+from types import BuiltinMethodType, TracebackType
 from typing import Any, Callable, Dict, Iterator, List, NoReturn, Optional, Tuple, Type
 
 from flexmock.exceptions import (
@@ -431,7 +431,7 @@ class Mock:
                         name, expectation._args
                     )
             # make sure to clean up expectations to ensure none of them
-            # interfere with the runner's error reporing mechanism
+            # interfere with the runner's error reporting mechanism
             # e.g. open()
             for _, expectations in FlexmockContainer.flexmock_objects.items():
                 for expectation in expectations:
@@ -560,26 +560,25 @@ class Expectation:
         if isinstance(self._mock, Mock):
             return  # no sense in enforcing this for fake objects
         allowed = self._argspec
-        # we consider object a method for purposes or not counting "self"/"cls" as argument if:
-        #  - one of inspect.ismethod, inspect.isfunction, _isclass return True
-        #    (in Python 3 it's sometimes impossible to tell whether callable is method or not,
-        #     so we try both inspect.ismethod and inspect.isfunction; classes are callable too -
-        #     they have __init__)
-        #  - it's not a static method
-        #  - the mocked object is a module - module "methods" are in fact plain functions;
-        #    unless they're classes, which means they still have __init__
-        is_method = (
-            (
-                inspect.ismethod(self._original)
-                or inspect.isfunction(self._original)
-                or inspect.isclass(self._original)
-            )
-            and self._method_type is not staticmethod
-            and (not isinstance(self._mock, types.ModuleType) or inspect.isclass(self._original))
-        )
         args_len = len(allowed.args)
-        if is_method:
+
+        # Builtin methods take `self` as the first argument but `inspect.ismethod` returns False
+        # so we need to check for them explicitly
+        is_builtin_method = isinstance(self._original, BuiltinMethodType)
+        # Methods take `self` if not a staticmethod
+        is_method = inspect.ismethod(self._original) and self._method_type is not staticmethod
+        # Class init takes `self`
+        is_class = inspect.isclass(self._original)
+        # When calling class methods or instance methods on a class method takes `cls`
+        is_class_method = (
+            inspect.isfunction(self._original)
+            and inspect.isclass(self._mock)
+            and self._method_type is not staticmethod
+        )
+        if is_builtin_method or is_method or is_class or is_class_method:
+            # Do not count `self` or `cls`.
             args_len -= 1
+
         minimum = args_len - (allowed.defaults and len(allowed.defaults) or 0)
         maximum = None
         if allowed.varargs is None and allowed.varkw is None:

--- a/tests/flexmock_test.py
+++ b/tests/flexmock_test.py
@@ -1,5 +1,6 @@
 """Flexmock tests."""
 # pylint: disable=missing-docstring,too-many-lines,disallowed-name,no-member,invalid-name,no-self-use
+import random
 import re
 import sys
 import unittest
@@ -312,6 +313,15 @@ class RegularClass:
         assert_equal("got a string", mock.method_foo("string!"))
         assert_equal("got an int", mock.method_foo(23))
         assert_raises(MethodSignatureError, mock.method_foo, 2.0)
+
+    def test_with_args_should_work_with_builtin_c_functions_and_methods(self):
+        flexmock(sys.stdout).should_call("write")  # set fall-through
+        flexmock(sys.stdout).should_receive("write").with_args("flexmock_builtin_test").once()
+        sys.stdout.write("flexmock_builtin_test")
+
+    def test_with_args_should_work_with_builtin_python_methods(self):
+        flexmock(random).should_receive("randint").with_args(1, 10).once()
+        random.randint(1, 10)
 
     def test_flexmock_should_match_expectations_against_user_defined_classes(self):
         mock = flexmock(name="temp")

